File: tools/my_devices/oven.py
from __future__ import annotations

from ..base import BaseTool, ToolResult
from ..schemas.common import GetDeviceStatusInput


class OvenStatusTool(BaseTool):
    name = "oven_status"
    description = "오븐/쿠커의 현재 상태를 조회합니다. 운전 모드, 동작 상태, 온도 등을 확인할 수 있습니다. (읽기 전용)"
    args_schema = GetDeviceStatusInput

    def execute(self, args: GetDeviceStatusInput) -> ToolResult:
        response = self.client.get(f"/devices/{args.device_id}/status")
        if isinstance(response, dict) and "error" in response:
            return ToolResult(
                success=False,
                error=response["error"].get("message", "Failed to get oven status"),
            )

        main = (
            response.get("components", {}).get("main", {})
            if isinstance(response, dict)
            else {}
        )

        oven_mode = main.get("ovenMode", {}).get("ovenMode", {}).get("value")
        oven_operating_state = (
            main.get("ovenOperatingState", {}).get("machineState", {}).get("value")
        )
        temperature_value = (
            main.get("temperatureMeasurement", {}).get("temperature", {}).get("value")
        )
        temperature_unit = (
            main.get("temperatureMeasurement", {}).get("temperature", {}).get("unit")
        )

        data: dict[str, object] = {
            "oven_mode": oven_mode,
            "oven_operating_state": oven_operating_state,
            "temperature": {"value": temperature_value, "unit": temperature_unit},
        }

        return ToolResult(success=True, data=data)

File: smartthings_tools_rs/src/my_devices/oven.rs
use schemars::schema::RootSchema;
use schemars::schema_for;
use serde_json::Value;

use crate::client::SmartThingsClient;
use crate::common_tools::has_error;
use crate::schemas::my_devices::OvenStatusInput;
use crate::tool::{Tool, ToolResult};

pub struct OvenStatusTool<'a> {
    pub client: &'a SmartThingsClient,
}

impl Tool for OvenStatusTool<'_> {
    fn name(&self) -> &'static str {
        "oven_status"
    }
    fn description(&self) -> &'static str {
        "오븐/쿠커의 현재 상태를 조회합니다. 운전 모드, 동작 상태, 온도 등을 확인할 수 있습니다. (읽기 전용)"
    }
    fn parameters_schema(&self) -> RootSchema {
        schema_for!(OvenStatusInput)
    }

    fn execute(&self, args: Value) -> ToolResult {
        let input: OvenStatusInput = match serde_json::from_value(args) {
            Ok(v) => v,
            Err(e) => return ToolResult::fail(format!("ValidationError: {e}")),
        };

        let response = self
            .client
            .get(&format!("/devices/{}/status", input.device_id), None);
        if has_error(&response) {
            return ToolResult::fail(crate::common_tools::extract_error_msg(&response));
        }

        let main = response.get("components").and_then(|c| c.get("main"));

        let oven_mode = main
            .and_then(|m| m.get("ovenMode"))
            .and_then(|o| o.get("ovenMode"))
            .and_then(|o| o.get("value"));

        let operating_state = main
            .and_then(|m| m.get("ovenOperatingState"))
            .and_then(|o| o.get("machineState"))
            .and_then(|o| o.get("value"));

        let temp_value = main
            .and_then(|m| m.get("temperatureMeasurement"))
            .and_then(|t| t.get("temperature"))
            .and_then(|t| t.get("value"));

        let temp_unit = main
            .and_then(|m| m.get("temperatureMeasurement"))
            .and_then(|t| t.get("temperature"))
            .and_then(|t| t.get("unit"));

        ToolResult::ok(serde_json::json!({
            "oven_mode": oven_mode,
            "oven_operating_state": operating_state,
            "temperature": { "value": temp_value, "unit": temp_unit },
        }))
    }
}
